### PR TITLE
Initial support for reading decimal types from JSON and CSV

### PR DESCRIPTION
**Pull request by andygrove**
Signed-off-by: Andy Grove <andygrove@nvidia.com>

Closes https://github.com/NVIDIA/spark-rapids/issues/4824 and https://github.com/NVIDIA/spark-rapids/issues/4615 and https://github.com/NVIDIA/spark-rapids/issues/4646

The main change in this PR is adding support for reading decimal values from CSV and JSON by reading them as strings and then casting them to decimal.

There is also a fix for a bug in handling invalid JSON floating-point values such as `.1` and `1.`which was discovered during this work.

There are also some updates to the compatibility guide related to previous JSON and CSV PRs. I can PR those docs changes separately if that is preferred.

### CODE REVIEW

File: docs/compatibility.md
Comment (andygrove) [inline docs/compatibility.md:287]: These changes are not specific to this PR and could be PR'd separately

File: integration_tests/src/test/resources/decimals.csv
Comment (revans2) [inline integration_tests/src/test/resources/decimals.csv:12]: Could we have a few more numbers to see if Spark is rounding when the scale is larger than expected, or if it is truncating?

File: integration_tests/src/test/resources/decimals.json
Comment (revans2) [inline integration_tests/src/test/resources/decimals.json:11]: Same here it would be nice to have some numbers that would possibly round down instead of up if Spark was rounding.

File: sql-plugin/src/main/scala/org/apache/spark/sql/catalyst/json/rapids/GpuJsonScan.scala
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/catalyst/json/rapids/GpuJsonScan.scala:374]: I think `it's` needs to be `its`

File: sql-plugin/src/main/scala/org/apache/spark/sql/catalyst/json/rapids/GpuJsonScan.scala
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/catalyst/json/rapids/GpuJsonScan.scala:392]: Is there any way we can do this without a regular expression?  If not could we file a follow on issue to figure out how to speed this up?

Comment (andygrove) [reply sql-plugin/src/main/scala/org/apache/spark/sql/catalyst/json/rapids/GpuJsonScan.scala:392]: Yes, I can rework this to avoid regexp. I will do so in this PR.

**Review by revans2 — APPROVED** [review]
Looking good.

File: sql-plugin/src/main/scala/org/apache/spark/sql/catalyst/json/rapids/GpuJsonScan.scala
Comment (andygrove) [reply sql-plugin/src/main/scala/org/apache/spark/sql/catalyst/json/rapids/GpuJsonScan.scala]: I simplified this a bit but it is still calling into `GpuCast` which applies additional complex regular expressions that are mostly redundant because we have already sanitized the input here, so I have filed https://github.com/NVIDIA/spark-rapids/issues/4837 to optimize this farther

**Comment by andygrove** [comment]
build

**Comment by andygrove** [comment]
@revans2 Could I get another review on this one? There were changes after your last approval, to reduce the amount of regexp in use.

File: sql-plugin/src/main/scala/org/apache/spark/sql/catalyst/json/rapids/GpuJsonScan.scala
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/catalyst/json/rapids/GpuJsonScan.scala:384]: nit: can you file a follow on issue in the spark rapids repo to look into ways to remove the regular expression? I am hopeful it will be fixed as a part of the rewrite of JSON parsing, but if not I want us to track it.

File: sql-plugin/src/main/scala/org/apache/spark/sql/catalyst/json/rapids/GpuJsonScan.scala
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/catalyst/json/rapids/GpuJsonScan.scala:411]: nit: Decimal does not support NaN or Inf, etc. It would be nice to not have to to the extra checks if we are parsing decimal, but allowNonNumbericNumbers happens to be set.

**Review by revans2 — APPROVED** [review]
Just two nits related to performance that I can live without.

File: sql-plugin/src/main/scala/org/apache/spark/sql/catalyst/json/rapids/GpuJsonScan.scala
Comment (revans2) [reply sql-plugin/src/main/scala/org/apache/spark/sql/catalyst/json/rapids/GpuJsonScan.scala:384]: Sorry you already did this one :)

**Pull request merged**
